File: backend/app/models/widget.py
"""
Widget Model

Defines the data structure for widgets - floating overlays that can display
live streams or embedded content on any page.

Widget Types:
- system: Created by admins, visible to all applicable users
- personal: Created by individual users, visible only to them
"""

from datetime import datetime
from typing import Optional, List
from enum import Enum
from beanie import Document
from pydantic import BaseModel, Field


class WidgetType(str, Enum):
    """Type of widget ownership"""
    SYSTEM = "system"      # Admin-created, visible to targeted users
    PERSONAL = "personal"  # User-created, visible only to owner


class WidgetContentType(str, Enum):
    """Type of content displayed in widget"""
    LIVE_CHANNEL = "live_channel"  # Live TV stream
    IFRAME = "iframe"              # External embed URL
    PODCAST = "podcast"            # Podcast episode
    VOD = "vod"                    # Video on demand
    RADIO = "radio"                # Radio station
    LIVE = "live"                  # Generic live content
    CUSTOM = "custom"              # Custom React component


class WidgetPosition(BaseModel):
    """Position and size of widget on screen"""
    x: float = 20.0           # Pixels from left (or right in RTL)
    y: float = 100.0          # Pixels from top
    width: int = 320          # Widget width in pixels
    height: int = 180         # Widget height in pixels (16:9 ratio default)
    z_index: int = 100        # Layering order


class WidgetContent(BaseModel):
    """Content configuration for widget"""
    content_type: WidgetContentType

    # For LIVE_CHANNEL type
    live_channel_id: Optional[str] = None

    # For PODCAST, VOD, RADIO, LIVE types
    podcast_id: Optional[str] = None
    content_id: Optional[str] = None  # Generic content ID for VOD
    station_id: Optional[str] = None

    # For IFRAME type
    iframe_url: Optional[str] = None
    iframe_title: Optional[str] = None

    # For CUSTOM type (React components)
    component_name: Optional[str] = None  # e.g., "ynet_mivzakim"


class Widget(Document):
    """
    Widget document for MongoDB storage.

    Widgets are floating overlay components that can display live streams
    or embedded content anywhere on the application.
    """

    # Identity
    type: WidgetType                          # system or personal
    user_id: Optional[str] = None             # Only for personal widgets (owner)

    # Display
    title: str
    description: Optional[str] = None
    icon: Optional[str] = None                # Icon URL or emoji

    # Content
    content: WidgetContent

    # Positioning (default values, can be overridden per-user)
    position: WidgetPosition = Field(default_factory=WidgetPosition)

    # Behavior flags
    is_active: bool = True                    # Whether widget is enabled
    is_muted: bool = True                     # Default mute state (all muted by default)
    is_visible: bool = True                   # Default visibility state
    is_closable: bool = True                  # Allow users to close
    is_draggable: bool = True                 # Allow users to reposition

    # Targeting (for system widgets)
    visible_to_roles: List[str] = Field(default_factory=lambda: ["user"])
    visible_to_subscription_tiers: List[str] = Field(default_factory=list)
    target_pages: List[str] = Field(default_factory=list)  # Empty = all pages

    # Ordering
    order: int = 0                            # Display order when multiple widgets

    # Metadata
    created_by: Optional[str] = None          # Admin user ID for system widgets
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "widgets"
        indexes = [
            "type",
            "user_id",
            "is_active",
            ("type", "is_active"),
            ("user_id", "is_active"),
            "created_at",
            "order",
        ]


# Pydantic schemas for API requests/responses

class WidgetCreateRequest(BaseModel):
    """Request schema for creating a widget"""
    title: str
    description: Optional[str] = None
    icon: Optional[str] = None
    content: WidgetContent
    position: Optional[WidgetPosition] = None
    is_muted: bool = True
    is_closable: bool = True
    is_draggable: bool = True
    # Targeting (only for system widgets)
    visible_to_roles: List[str] = Field(default_factory=lambda: ["user"])
    visible_to_subscription_tiers: List[str] = Field(default_factory=list)
    target_pages: List[str] = Field(default_factory=list)
    order: int = 0


class WidgetUpdateRequest(BaseModel):
    """Request schema for updating a widget (all fields optional)"""
    title: Optional[str] = None
    description: Optional[str] = None
    icon: Optional[str] = None
    content: Optional[WidgetContent] = None
    position: Optional[WidgetPosition] = None
    is_active: Optional[bool] = None
    is_muted: Optional[bool] = None
    is_visible: Optional[bool] = None
    is_closable: Optional[bool] = None
    is_draggable: Optional[bool] = None
    visible_to_roles: Optional[List[str]] = None
    visible_to_subscription_tiers: Optional[List[str]] = None
    target_pages: Optional[List[str]] = None
    order: Optional[int] = None


class WidgetPositionUpdate(BaseModel):
    """Request schema for updating widget position only"""
    x: float
    y: float
    width: Optional[int] = None
    height: Optional[int] = None


class WidgetResponse(BaseModel):
    """Response schema for widget data"""
    id: str
    type: WidgetType
    user_id: Optional[str] = None
    title: str
    description: Optional[str] = None
    icon: Optional[str] = None
    content: WidgetContent
    position: WidgetPosition
    is_active: bool
    is_muted: bool
    is_visible: bool
    is_closable: bool
    is_draggable: bool
    visible_to_roles: List[str]
    visible_to_subscription_tiers: List[str]
    target_pages: List[str]
    order: int
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: olorin-media/bayit-plus/web/src/types/widget.ts
/**
 * Widget Types
 *
 * TypeScript interfaces for widget system - floating overlays
 * that display live streams or embedded content.
 */

export type WidgetType = 'system' | 'personal';

export type WidgetContentType = 'live_channel' | 'iframe' | 'podcast' | 'vod' | 'radio' | 'live' | 'custom';

export interface WidgetPosition {
  x: number;
  y: number;
  width: number;
  height: number;
  z_index: number;
}

export interface WidgetContent {
  content_type: WidgetContentType;
  live_channel_id?: string | null;
  podcast_id?: string | null;
  content_id?: string | null;
  station_id?: string | null;
  iframe_url?: string | null;
  iframe_title?: string | null;
  component_name?: string | null;  // For custom React components
}

export interface Widget {
  id: string;
  type: WidgetType;
  user_id?: string | null;
  title: string;
  description?: string | null;
  icon?: string | null;
  cover_url?: string | null;
  content: WidgetContent;
  position: WidgetPosition;
  is_active: boolean;
  is_muted: boolean;
  is_visible: boolean;
  is_minimized: boolean;
  is_closable: boolean;
  is_draggable: boolean;
  visible_to_roles: string[];
  visible_to_subscription_tiers: string[];
  target_pages: string[];
  order: number;
  created_by?: string | null;
  created_at: string;
  updated_at: string;
}

// Request types

export interface WidgetCreateRequest {
  title: string;
  description?: string;
  icon?: string;
  content: WidgetContent;
  position?: Partial<WidgetPosition>;
  is_muted?: boolean;
  is_closable?: boolean;
  is_draggable?: boolean;
  visible_to_roles?: string[];
  visible_to_subscription_tiers?: string[];
  target_pages?: string[];
  order?: number;
}

export interface WidgetUpdateRequest {
  title?: string;
  description?: string | null;
  icon?: string | null;
  content?: WidgetContent;
  position?: Partial<WidgetPosition>;
  is_active?: boolean;
  is_muted?: boolean;
  is_visible?: boolean;
  is_minimized?: boolean;
  is_closable?: boolean;
  is_draggable?: boolean;
  visible_to_roles?: string[];
  visible_to_subscription_tiers?: string[];
  target_pages?: string[];
  order?: number;
}

export interface WidgetPositionUpdate {
  x: number;
  y: number;
  width?: number;
  height?: number;
}

// Response types

export interface WidgetsListResponse {
  items: Widget[];
  total: number;
  page?: number;
  page_size?: number;
  total_pages?: number;
}

// Client-side state (extends backend widget with local state)

export interface WidgetClientState {
  isMuted: boolean;
  isVisible: boolean;
  isMinimized: boolean;
  position: WidgetPosition;
}

export interface WidgetWithState extends Widget {
  clientState: WidgetClientState;
}

// Form types for admin UI

export interface WidgetFormData {
  title: string;
  description: string;
  icon: string;
  content_type: WidgetContentType;
  content_id: string;  // Unified content ID (maps to live_channel_id, podcast_id, content_id, or station_id)
  podcast_id?: string;
  station_id?: string;
  iframe_url: string;
  iframe_title: string;
  position_x: number;
  position_y: number;
  position_width: number;
  position_height: number;
  is_muted: boolean;
  is_closable: boolean;
  is_draggable: boolean;
  visible_to_roles: string[];
  visible_to_subscription_tiers: string[];
  target_pages: string[];
  order: number;
}

// Helper to convert form data to API request
export function formDataToCreateRequest(data: WidgetFormData): WidgetCreateRequest {
  return {
    title: data.title,
    description: data.description || undefined,
    icon: data.icon || undefined,
    content: {
      content_type: data.content_type,
      live_channel_id: data.content_type === 'live_channel' || data.content_type === 'live' ? data.content_id : null,
      podcast_id: data.content_type === 'podcast' ? data.content_id : null,
      content_id: data.content_type === 'vod' ? data.content_id : null,
      station_id: data.content_type === 'radio' ? data.content_id : null,
      iframe_url: data.content_type === 'iframe' ? data.iframe_url : null,
      iframe_title: data.content_type === 'iframe' ? data.iframe_title : null,
    },
    position: {
      x: data.position_x,
      y: data.position_y,
      width: data.position_width,
      height: data.position_height,
    },
    is_muted: data.is_muted,
    is_closable: data.is_closable,
    is_draggable: data.is_draggable,
    visible_to_roles: data.visible_to_roles,
    visible_to_subscription_tiers: data.visible_to_subscription_tiers,
    target_pages: data.target_pages,
    order: data.order,
  };
}

// Default position for new widgets
export const DEFAULT_WIDGET_POSITION: WidgetPosition = {
  x: 20,
  y: 100,
  width: 640,
  height: 360,
  z_index: 100,
};

// User System Widget - tracks user's subscription to a system widget
export interface UserSystemWidget {
  id: string;
  user_id: string;
  widget_id: string;
  added_at: string;
  position?: WidgetPosition;
  is_muted: boolean;
  is_visible: boolean;
  is_minimized: boolean;
  order: number;
}

// Available system widget with subscription status
export interface AvailableSystemWidget extends Widget {
  is_added: boolean;
}

// Available system widgets response
export interface AvailableSystemWidgetsResponse {
  items: AvailableSystemWidget[];
  total: number;
}

// Default form values
export const DEFAULT_WIDGET_FORM: WidgetFormData = {
  title: '',
  description: '',
  icon: '',
  content_type: 'live_channel',
  content_id: '',
  iframe_url: '',
  iframe_title: '',
  position_x: DEFAULT_WIDGET_POSITION.x,
  position_y: DEFAULT_WIDGET_POSITION.y,
  position_width: DEFAULT_WIDGET_POSITION.width,
  position_height: DEFAULT_WIDGET_POSITION.height,
  is_muted: true,
  is_closable: true,
  is_draggable: true,
  visible_to_roles: ['user'],
  visible_to_subscription_tiers: [],
  target_pages: [],
  order: 0,
};
